File: Packages/js-common/Source/DB/nodeLinks/NodeLinkValidity.ts
import {CreateAccessor} from "web-vcore/nm/mobx-graphlink.js";
import {GetValues} from "web-vcore/nm/js-vextensions.js";
import {globalRootNodeID} from "../../DB_Constants.js";
import {DoesPolicyAllowX} from "../@Shared/TablePermissions.js";
import {APAction, APTable} from "../accessPolicies/@PermissionSet.js";
import {GetNodeLinks} from "../nodeLinks.js";
import {NodeL1, NodeL2} from "../nodes/@Node.js";
import {NodeType, NodeType_Info} from "../nodes/@NodeType.js";
import {CanGetBasicPermissions, GetUserPermissionGroups, HasAdminPermissions} from "../users/$user.js";
import {User} from "../users/@User.js";
import {GetNode} from "../nodes.js";
import {ChildGroup, Polarity, childGroupsWithPolarity_required, childGroupsWithPolarity_requiredOrOptional} from "./@NodeLink.js";
import {PickOnly} from "../../Utils.js";

/*export function GetValidChildTypes(nodeType: NodeType, path: string, group: ChildGroup, polarity: Polarity|n) {
	const nodeTypes = GetValues(NodeType);
	const validChildTypes = nodeTypes.filter(type=>CheckLinkIsValid(nodeType, type, group, polarity) == null);
	return validChildTypes;
}
export function GetValidNewChildTypes(parent: NodeL2, childGroup: ChildGroup, polarity: Polarity|n, actor: User|n) {
	const nodeTypes = GetValues(NodeType);
	const validChildTypes = nodeTypes.filter(type=>CheckNewLinkIsValid(parent.id, {type} as any, childGroup, polarity, actor) == null);

	// if in relevance or truth group, claims cannot be direct children (must be within argument)
	/*if (childGroup == ChildGroup.relevance || childGroup == ChildGroup.truth) {
		validChildTypes = validChildTypes.Exclude(NodeType.claim);
	} else {
		// in the other cases, arguments cannot be direct children (those are only meant for in relevance/truth groups)
		validChildTypes = validChildTypes.Exclude(NodeType.argument);
	}*#/

	return validChildTypes;
}*/

export class NewChildConfig {
	constructor(data: Omit<NewChildConfig, "FinalChildType">) {
		Object.assign(this, data);
	}
	childGroup: ChildGroup;
	/** Note: If addWrapperArg is true, this actually represents the grandchild-type, ie. type of node being put in the wrapper-argument. */
	childType: NodeType;
	addWrapperArg: boolean;
	/** Note: If addWrapperArg is true, this represents the polarity of the wrapper-argument, rather than than the grandchild. */
	polarity: Polarity|n;

	FinalChildType() {
		return this.addWrapperArg ? NodeType.argument : this.childType;
	}
}
function GetAllNewChildConfigs() {
	const result = [] as NewChildConfig[];
	const nodeTypes = GetValues(NodeType);
	for (const childGroup of GetValues(ChildGroup)) {
		for (const childType of nodeTypes) {
			// filtering based on this array is merely an optimization
			/*const addWrapperArgPossibilities = [false];
			if (polarity != null && childType == NodeType.claim) addWrapperArgPossibilities.push(true);*/
			const addWrapperArgPossibilities = [false, true];

			for (const addWrapperArg of addWrapperArgPossibilities) {
				// filtering based on this array is merely an optimization
				/*const polarityPossibilities: Array<Polarity|n> = [null];
				if (childGroupsWithPolarity_requiredOrOptional.includes(childGroup)) polarityPossibilities.push(Polarity.supporting, Polarity.opposing);*/
				const polarityPossibilities = [null, Polarity.supporting, Polarity.opposing];

				for (const polarity of polarityPossibilities) {
					result.push(new NewChildConfig({childType, childGroup, polarity, addWrapperArg}));
				}
			}
		}
	}
	return result;
}
export const allNewChildConfigs = GetAllNewChildConfigs();

/*export function GetPossiblyValidNewChildConfigs(parentType: NodeType) {
	return allNewChildConfigs.filter(config=>CheckLinkIsValid(parentType, config.childType, config.childGroup, config.polarity) == null);
}*/

export function CheckNewChildConfigUnderParentIsValid(config: NewChildConfig, parentID: string, actor: User|n) {
	// `addWrapperArg` represents functionality beyond what CheckNewLinkIsValid can check; so do the validation for that extra functionality here
	// (more elegant way would be to enable a call to CheckNewLinkIsValid for the wrapperArg->claim link as well, but not possible atm since there's no parent-id resolvable for the wrapper-arg)
	if (config.addWrapperArg) {
		if (config.childType != NodeType.claim) return "Only claims can have wrapper-arguments added for them.";
	}
	return CheckNewLinkIsValid(parentID, null, config.FinalChildType(), config.childGroup, config.polarity, actor);
}
export function GetValidNewChildConfigsUnderParent(parentID: string, actor: User|n) {
	return allNewChildConfigs.filter(config=>CheckNewChildConfigUnderParentIsValid(config, parentID, actor) == null);
}

/** Does basic checking of validity of parent<>child linkage. See `CheckValidityOfNewLink` for a more thorough validation. */
// sync: rs[assert_link_is_valid]
export const CheckLinkIsValid = CreateAccessor((parentType: NodeType, childType: NodeType, childGroup: ChildGroup, linkPolarity: Polarity|n)=>{
	// redundant check, improving error-message clarity for certain issues
	if (!NodeType_Info.for[parentType].childGroup_childTypes.has(childGroup)) {
		return `Where parent's type is ${NodeType[parentType]}, no "${ChildGroup[childGroup]}" child-group exists.`;
	}

	const validChildTypes = NodeType_Info.for[parentType].childGroup_childTypes.get(childGroup) ?? [];
	if (!validChildTypes.includes(childType)) {
		// redundant checks, improving error-message clarity for certain issues
		if (parentType == NodeType.argument && childGroup == ChildGroup.generic && childType != NodeType.claim) {
			return `Where parent is an argument, and child-group is generic, a claim child is expected (instead it's a ${NodeType[childType]}).`;
		}
		const isSLSimpleArg = parentType == NodeType.claim && childType == NodeType.claim && childGroup == ChildGroup.truth && linkPolarity != null;
		if (!isSLSimpleArg && childGroupsWithPolarity_required.includes(childGroup) && childType != NodeType.argument) {
			return `Where child-group is ${childGroup}, an argument child is expected (instead it's a ${NodeType[childType]}).`;
		}

		// give generic message
		return `The child's type (${NodeType[childType]}) is not valid here. (parent type: ${NodeType[parentType]}, child group: ${ChildGroup[childGroup]})`;
	}

	if (childGroupsWithPolarity_required.includes(childGroup)) {
		if (linkPolarity == null) return `A link in group "truth", "relevance", or "neutrality" must have a polarity specified.`;
	}
	if (childType == NodeType.argument) {
		if (linkPolarity == null) return "A link with an argument child must have a polarity specified.";
	}
	if (linkPolarity != null) {
		if (!childGroupsWithPolarity_requiredOrOptional.includes(childGroup)) return `Only links in child-group "truth", "relevance", "neutrality", or "freeform" can have a polarity specified.`;
		if (!(childType == NodeType.argument || childType == NodeType.claim)) return "Only links with an argument child (or claim child, in sl-mode) can have a polarity specified.";
	}
});
/**
 * Extension of `CheckValidityOfLink`, with additional checking based on knowledge of specific nodes being linked, user's permissions, etc.
 * For example:
 * * Blocks if node is being linked as child of itself.
 * * Blocks if adding child to global-root, without user being an admin.
 * */
// sync: rs[assert_new_link_is_valid]
export const CheckNewLinkIsValid = CreateAccessor((parentID: string, newChildID: string|null, newChildType: NodeType, newChildGroup: ChildGroup, newLinkPolarity: Polarity|n, actor: User|n)=>{
	const permissions = GetUserPermissionGroups(actor?.id);
	if (!CanGetBasicPermissions(permissions)) return "You're not signed in, or lack basic permissions.";

	const parent = GetNode(parentID);
	if (parent == null) return "Parent data not found.";
	//if (!) { return "Parent node's permission policy does not grant you the ability to add children."; }
	const guessedCanAddChild = actor
		? NodeL1.canAddChild(parent, actor) // if can add child
		: DoesPolicyAllowX(null, parent.accessPolicy, APTable.nodes, APAction.addChild); // or probably can
	if (!guessedCanAddChild) { return "Parent node's permission policy does not grant you the ability to add children."; }

	// const parentPathIDs = SplitStringBySlash_Cached(parentPath).map(a => a.ToInt());
	// if (map.name == "Global" && parentPathIDs.length == 1) return false; // if parent is l1(root), don't accept new children
	if (parent.id == globalRootNodeID && !HasAdminPermissions(permissions)) return "Only admins can add children to the global-root.";
	// if in global map, parent is l2, and user is not a mod (and not node creator), don't accept new children
	// if (parentPathIDs[0] == globalRootNodeID && parentPathIDs.length == 2 && !HasModPermissions(permissions) && parent.creator != MeID()) return false;
	if (parent.id == newChildID) return "Cannot link node as its own child.";

	const parentChildLinks = GetNodeLinks(parentID, null, newChildGroup); // query it with "childID" null, so it's cached once for all such calls
	const isAlreadyChild = parentChildLinks.Any(a=>a.child == newChildID);

	// if new-holder-type is not specified, consider "any" and so don't check
	/*if (newChildGroup !== undefined) {
		const currentChildGroup = GetChildGroup(newChild.type, parent.type);
		if (isAlreadyChild && currentChildGroup == newChildGroup) return "Node is already a child of the parent."; // if already a child of this parent, reject (unless it's a claim, in which case allow, as can be)
	}*/
	if (isAlreadyChild) return "Node is already a child of the parent.";

	return CheckLinkIsValid(parent.type, newChildType, newChildGroup, newLinkPolarity);
});

File: Packages/app-server/src/db/node_links_/node_link_validity.rs
use rust_shared::async_graphql::{ID, SimpleObject, InputObject};
use rust_shared::rust_macros::wrap_slow_macros;
use rust_shared::serde_json::{Value, json};
use rust_shared::db_constants::{SYSTEM_USER_ID, GLOBAL_ROOT_NODE_ID};
use rust_shared::{async_graphql, serde_json, anyhow, GQLError};
use rust_shared::async_graphql::{Object};
use rust_shared::utils::type_aliases::JSONValue;
use rust_shared::anyhow::{anyhow, Error, ensure, bail, Context};
use rust_shared::utils::time::{time_since_epoch_ms_i64};
use rust_shared::serde::{Serialize, Deserialize};
use tracing::info;

use crate::db::_shared::access_policy_target::AccessPolicyTarget;
use crate::db::_shared::common_errors::err_should_be_populated;
use crate::db::_shared::table_permissions::{does_policy_allow_x, CanVote, CanAddChild};
use crate::db::access_policies::get_access_policy;
use crate::db::access_policies_::_permission_set::{APAction, APTable};
use crate::db::commands::_command::command_boilerplate;
use crate::db::general::permission_helpers::assert_user_can_add_child;
use crate::db::general::sign_in_::jwt_utils::{resolve_jwt_to_user_info, get_user_info_from_gql_ctx};
use crate::db::node_links::{get_node_links, ChildGroup, NodeLink, NodeLinkInput, Polarity, CHILD_GROUPS_WITH_POLARITY_REQUIRED_OR_OPTIONAL, CHILD_GROUPS_WITH_POLARITY_REQUIRED};
use crate::db::nodes::get_node;
use crate::db::nodes_::_node::{Node};
use crate::db::nodes_::_node_type::{get_node_type_info, NodeType};
use crate::db::users::{User, PermissionGroups, get_user};
use crate::utils::db::accessors::AccessorContext;
use rust_shared::utils::db::uuid::new_uuid_v4_as_b64;
use crate::utils::general::data_anchor::{DataAnchorFor1};

/// Does basic checking of validity of parent<>child linkage. See `assert_new_link_is_valid` for a more thorough validation.
// sync: js[CheckLinkIsValid]
pub fn assert_link_is_valid(parent_type: NodeType, child_type: NodeType, child_group: ChildGroup, link_polarity: Option<Polarity>) -> Result<(), Error> {
	// redundant check, improving error-message clarity for certain issues
	if !get_node_type_info(parent_type).childGroup_childTypes.contains_key(&child_group) {
		bail!("Where parent's type is {parent_type:?}, no \"{child_group:?}\" child-group exists.");
	}

	let valid_child_types = get_node_type_info(parent_type).childGroup_childTypes.get(&child_group).cloned().unwrap_or(vec![]);
	if !valid_child_types.contains(&child_type) {
		// redundant checks, improving error-message clarity for certain issues
		if parent_type == NodeType::argument && child_group == ChildGroup::generic && child_type != NodeType::claim {
			bail!("Where parent is an argument, and child-group is generic, a claim child is expected (instead it's a {child_type:?}).");
		}
		let is_sl_simple_arg = parent_type == NodeType::claim && child_type == NodeType::claim && child_group == ChildGroup::truth && link_polarity.is_some();
		if !is_sl_simple_arg && CHILD_GROUPS_WITH_POLARITY_REQUIRED.contains(&child_group) && child_type != NodeType::argument {
			bail!("Where child-group is {child_group:?}, an argument child is expected (instead it's a {child_type:?}).");
		}

		// give generic message
		bail!("The child's type ({child_type:?}) is not valid here. (parent type: {parent_type:?}, child group: ${child_group:?})");
	}

	if CHILD_GROUPS_WITH_POLARITY_REQUIRED.contains(&child_group) {
		ensure!(link_polarity.is_some(), r#"A link in group "truth", "relevance", or "neutrality" must have a polarity specified."#);
	}
	if child_type == NodeType::argument {
		ensure!(link_polarity.is_some(), "A link with an argument child must have a polarity specified.");
	}
	if link_polarity.is_some() {
        ensure!(CHILD_GROUPS_WITH_POLARITY_REQUIRED_OR_OPTIONAL.contains(&child_group), r#"Only links in child-groups "truth", "relevance", "neutrality", or "freeform" can have a polarity specified."#);
		ensure!(child_type == NodeType::argument || child_type == NodeType::claim, "Only links with an argument child (or claim child, in sl-mode) can have a polarity specified.");
	}

	Ok(())
}

/// Extension of `assert_link_is_valid`, with additional checking based on knowledge of specific nodes being linked, user's permissions, etc.
/// For example:
/// * Blocks if node is being linked as child of itself.
/// * Blocks if adding child to global-root, without user being an admin.
// sync: js[CheckNewLinkIsValid]
pub async fn assert_new_link_is_valid(ctx: &AccessorContext<'_>, parent_id: &str, new_child_id: &str, new_child_type: NodeType, new_child_group: ChildGroup, new_link_polarity: Option<Polarity>, actor: Option<&User>) -> Result<(), Error> {
	// client-side version
	/*let permissions = if let Some(actor) = actor { actor.permissionGroups } else { PermissionGroups::all_false() };
	if !can_get_basic_permissions(actor, permissions) { bail!("You're not signed in, or lack basic permissions."); }*/
	// server-side version (on server, no need for this call-path to be called without actor)
	let (actor, permissions) = match actor {
		Some(actor) => (actor, &actor.permissionGroups),
		None => bail!("You're not signed in."),
	};
	if !permissions.basic { bail!("You lack basic permissions."); }
	
	let parent = get_node(ctx, parent_id).await.with_context(|| "Parent data not found")?;
	// client-side version
	//if !does_policy_allow_x(ctx, actor_id, &parent.accessPolicy, APTable::nodes, APAction::addChild).await? { bail!("Parent node's permission policy does not grant you the ability to add children."); }
	/*let guessedCanAddChild = match actor {
		Some(actor) => parent.can_add_child(ctx, actor).await?, // if can add child
		None => does_policy_allow_x(ctx, None, &parent.accessPolicy, APTable::nodes, APAction::addChild).await?, // or probably can
	};
	if !guessedCanAddChild { bail!("Parent node's permission policy does not grant you the ability to add children."); }*/
	// server-side version (on server, no need for this call-path to be called without actor)
	if !parent.can_add_child(ctx, actor).await? { bail!("Parent node's permission policy does not grant you the ability to add children."); }

	if parent.id == GLOBAL_ROOT_NODE_ID && !permissions.admin { bail!("Only admins can add children to the global-root."); }
	if parent.id == new_child_id { bail!("Cannot link node as its own child."); }

	let is_already_child = get_node_links(ctx, Some(parent_id), Some(&new_child_id)).await?.len() > 0;
	if is_already_child { bail!("Node is already a child of the parent."); }

	assert_link_is_valid(parent.r#type, new_child_type, new_child_group, new_link_polarity)
}